docs/references/ic-interface-spec.md

```diff
@@ -2202,6 +2223,67 @@ These system calls return costs in Cycles, represented by 128 bits, which will b
     - `1`: Invalid curve or algorithm. Memory at `dst` is left unchanged.
     - `2`: Invalid key name for the given combination of signing scheme and (valid) curve/algorithm. Memory at `dst` is left unchanged.
 
+### Environment Variables
+
+The following system calls provide access to the canister's environment variables:
+
+-   `ic0.env_var_count : () -> I`; `I ∈ {i32, i64}`
+
+    Returns the number of environment variables set for this canister.
+
+-   `ic0.env_var_name_size : (index: I) -> I`; `I ∈ {i32, i64}`
+
+    Gets the size in bytes of the name of the environment variable at the given index.
+
+    This system call traps if:
+      - If the index is out of bounds (>= than value provided by `ic0.env_var_count`)
+
+-   `ic0.env_var_name_copy : (index: I, dst: I, offset: I, size: I) -> ()`; `I ∈ {i32, i64}`
+
+    Copies the name of the environment variable at the given index into memory.
+
+    This system call traps if:
+      - The index is out of bounds (>= than value provided by `ic0.env_var_count`)
+      - `offset+size` is greater than the size of the environment variable name
+      - `dst+size` exceeds the size of the WebAssembly memory
+
+
+-   `ic0.env_var_name_exists : (name_src: I, name_size: I) -> i32`; `I ∈ {i32, i64}`
+
+    Checks if an environment variable with the given name exists. If yes, then a value of 1 is returned, otherwise a 0 is returned.
```

Comment:
I think this is a bit misleading. I would have expected to pass the actual name of an environment variable to `env_var_name_exists` reading the text, or is that encoded as `i32` / `i64` in `name_src`? In that case, why is `name_size` needed? in general i feel like i'm missing context on how the storage of the variables is implemented. i thought it's an array of byte arrays

Reply:
This is a low-level WASM system API that takes the name of the environment variable as input encoded by a pointer to the WASM memory `name_src` at which the name is stored (as UTF-8 encoded string) and the (binary) size `name_size` of the string. The result value is either 0 or 1 (encoded as `i32`). The low-level WASM system API cannot deal with high-level types (such as byte arrays) which is why they need to be represented by a pair of a pointer and size and this is how all existing system APIs are implemented.